US46 - Prerequisite Tasks

Daxxn opened the pull request:
### Overview

Reworks the prerequisite task system to look more in line with the rest of the app. Fixes bugs associated with over-complicated and duplicated code.

### Includes:

- client/src/main/clientData/UserData.ts
- client/src/main/components/PrereqTaskManager/PrereqProjectTaskList.tsx
- client/src/main/components/PrereqTaskManager/PrereqTaskDialog.tsx
- client/src/main/components/PrereqTaskManager/PrereqTaskList.tsx
- client/src/main/components/PrereqTaskManager/index.tsx
- client/src/main/components/ProjectTable/CompletableRow/PrereqTaskManager/PrereqList.tsx
- client/src/main/components/ProjectTable/CompletableRow/PrereqTaskManager/PrereqTaskDialog.tsx
- client/src/main/components/ProjectTable/CompletableRow/PrereqTaskManager/PrereqTaskDisplay.tsx
- client/src/main/components/ProjectTable/CompletableRow/PrereqTaskManager/index.tsx
- client/src/main/components/ProjectTable/CompletableRow/TaskMenu/index.tsx
- client/src/main/components/ProjectTable/CompletableRow/index.tsx
- client/src/main/components/ProjectTable/index.tsx
- client/src/main/utils/sortingFunctions.ts
- node_server/src/main/swagger.ts

### Developer Checklist:

- [x] My code compiles and runs locally
- [x] The code follows the `QualityPolicy.md` file
- [ ] My code has been developer-tested and includes unit tests
- [x] I have considered proper use of exceptions
- [x] I have eliminated IDE warnings

### Notes:

- Used the AutoComplete component to eliminate the need for a custom search solution.
- It was **UGLY!!**
- There's a bug that needs fixing : The AutoComplete clear button clears all selected prerequisites and not just the text in the text box. I'm not sure how to access the event to change that.
- Also added a few extra sorting functions that reverse the priority and prerequisite count functions.

Daxxn committed:
Fix missing PrereqTaskDialog in TaskMenu

Daxxn committed:
Merge remote-tracking branch 'origin/development' into US46-PrereqTasks-mergeDev

Daxxn committed:
Add PrereqTaskDisplay implementation to Completables

Daxxn committed:
Fixed bugs in PrereqTaskManager

A ) Modified the PrereqProjectTaskList to be able to add projects to the
prereq tasks.

B ) Added a new List for selected prereq tasks (AllPrereqTasks) to allow
separation of projects and tasks into 2. Allowing for better clarity of
the underlying functionality.

C ) Tried to fix some of the styling to pretty it up... it didnt work. I
may need to rewrite wome of the JSX.

D ) Added a sorting function to allow the projects to be sorted by
prerequisites.

E ) Other smaller refactors.

Daxxn committed:
Add swagger implementation for projects example

Daxxn committed:
WIP - Add Recursive loop to PrereqProjectTaskList

BUG - The resursion doesnt work yet. on the right track. will get it
put in later.

Daxxn committed:
WIP - Change how search results are displayed

Daxxn committed:
Remove useless code and ugly styling

Daxxn committed:
WIP - Add AutoComplete

Daxxn committed:
Fix prereqTask dialog & add AutoComplete search

Daxxn committed:
Remove uneeded components

Daxxn committed:
Fix cancel and save buttons on PrereqTaskDialog

I had to add a UserData listener to the PrereqTaskDialog to allow the
update of the Completable to occur in the PrereqTaskDialog. This also
removes the connection needed in the CompletableRow component.

Daxxn committed:
Fix alignment of buttons Add comments

Daxxn committed:
merged origin

Daxxn committed:
Merge branch 'development' into US46-prereqTasks-Auto

Had to move some things around and sme files were duped but it should be
working now.

Known Bug - The cancel button in the AutoComplete component clears all
the prerequisites when it should just clear the input text. I'm not sure
how to fix it. I couldn't find the property that controlls that action.

Daxxn committed:
Refactor property names & add comments

Daxxn committed:
Add typescript identifiers & small refactors

Daxxn committed:
Add Reversed sort functions

aneuhold commented:
Resolves #50

aneuhold committed:
I50 Fix vulnerabilities by resetting package-lock

aneuhold committed:
I50 Adjusted title of the PrereqTaskDialog

Adjusted the title so that it falls in line with the rest of the
application's dialog components. Also took out the font size setting so
that it globablly respects the user's font-size preference.

aneuhold reviewed (changes requested):
It looks like the user is allowed to add the task or project that is currently selected to the prereq tasks for that same task. 

When completing a task that is a prereq of another task, it doesn't seem that there are any visual changes that indicate that the prereq has been completed, or that the prereq task count has been decremented for the task that required it. This seems like it would require some extra logic on the server and on the client. Maybe a completed prereq could be crossed out when looking at the dialog?

Other than that, the functionality looks great! You even added the logic for cancelling changes, and that can be tough.

Daxxn commented:
I was trying to figure out how to update the visual indication of completion. I was having some issues with it. Ill take a look when i get off tonight.

aneuhold commented:
@Daxxn if you pull in the latest version of the `MobileUI` branch into this one, it might help a bit with UserData. `UserData` was split into separate classes for `User` and `Completables` but works exactly the same so it is easier to understand.

Daxxn committed:
Add Show Completed check box to Prerequisite Dialog

Possible bug:
duplication of tasks. Still not sure how it happened. I changed the
getOptionSelected property on the AutoComplete component when i noticed
some errors in the console saying there was components with duplicate
keys, looked and found 2 tasks duplicated. ONLY 2.
